File: source/com/barrybecker4/puzzle/sudoku1/ui/SudokuRenderer.java
// Copyright by Barry G. Becker, 2000-2011. Licensed under MIT License: http://www.opensource.org/licenses/MIT
package com.barrybecker4.puzzle.sudoku1.ui;

import com.barrybecker4.common.geometry.ByteLocation;
import com.barrybecker4.common.geometry.Location;
import com.barrybecker4.puzzle.sudoku1.model.ValueConverter;
import com.barrybecker4.puzzle.sudoku1.model.board.Board;
import com.barrybecker4.puzzle.sudoku1.model.board.Candidates;
import com.barrybecker4.puzzle.sudoku1.model.board.Cell;
import com.barrybecker4.ui.util.GUIUtil;

import java.awt.BasicStroke;
import java.awt.Color;
import java.awt.Font;
import java.awt.Graphics;
import java.awt.Graphics2D;
import java.awt.Point;
import java.awt.Stroke;
import java.util.Iterator;

/**
 * Renders the the sudoku puzzle onscreen.
 * @author Barry Becker
 */
public class SudokuRenderer implements CellLocator {

    private static final int MARGIN = 50;

    private static final Color CELL_ORIG_TEXT_COLOR = Color.BLACK;
    private static final Color CELL_TEXT_COLOR = Color.BLUE;

    private static final Color CELL_ORIG_BACKGROUND_COLOR = new Color(225, 225, 238, 200);
    private static final Color CELL_BACKGROUND_COLOR = new Color(245, 245, 255, 100);
    private static final Color CELL_FOCUS_COLOR = new Color(255, 250, 200);

    private static final Color USER_VALUE_COLOR = new Color(155, 5, 40);
    private static final Color USER_VALUE_CORRECT_COLOR = new Color(0, 200, 0);
    private static final Color USER_VALUE_WRONG_COLOR = new Color(255, 10, 0);
    private static final Color BIG_X_COLOR = new Color(255, 0, 0, 50);
    private static final Stroke BIG_X_STROKE = new BasicStroke(10, BasicStroke.CAP_BUTT, BasicStroke.JOIN_BEVEL);
    private static final Color GRID_COLOR = new Color(10, 0, 100);

    private static final Color TEXT_COLOR = new Color(0, 10, 10);
    private static final Color BACKGROUND_COLOR = new Color(220, 220, 240);
    private static final Color CANDIDATE_TEXT_COLOR = new Color(160, 160, 210);


    private static final Stroke CELL_STROKE = new BasicStroke(0.5f);
    private static final Stroke BIG_CELL_STROKE = new BasicStroke(3.0f);

    private Board board_;
    private boolean showCandidates = false;
    private int pieceSize;

    /**
     * Constructor
     */
    public SudokuRenderer(Board board) {
        board_ = board;
    }

    public void setShowCandidates(boolean show) {
         showCandidates = show;
    }

    public void setBoard(Board board) {
        board_ = board;
    }

    public Board getBoard() {
        return board_;
    }

    /**
     * This renders the current state of the Slider to the screen.
     */
    public void render(Graphics g, UserEnteredValues userEnteredValues, Location currentFocusLocation,
                       int width, int height)  {

        Graphics2D g2 = (Graphics2D) g;
        int minEdge = (Math.min(width, height) - 20 - MARGIN);
        pieceSize = minEdge / board_.getEdgeLength();
        // erase what's there and redraw.
        g.setColor( BACKGROUND_COLOR );
        g.fillRect( 0, 0, width, height );

        g.setColor( TEXT_COLOR );
        g.drawString( "Number of tries: " + board_.getNumIterations(),
                MARGIN, MARGIN - 24 );

        int len =  board_.getEdgeLength();
        int xpos, ypos;
        if (currentFocusLocation != null)  {
            drawCurrentFocus(g, currentFocusLocation);
        }

        for ( int i = 0; i < len; i++ ) {
            for ( int j = 0; j < len; j++ ) {
                Cell c = board_.getCell(i, j);

                xpos = MARGIN + j * pieceSize;
                ypos = MARGIN + i * pieceSize;

                drawCell(g2, c, xpos, ypos, userEnteredValues.get(i, j));
            }
        }
        drawCellBoundaryGrid(g, len);
    }

    public Location getCellCoordinates(Point point) {

        int row = (int)((point.getY() - MARGIN) / pieceSize);
        int col = (int)((point.getX() - MARGIN) / pieceSize);
        return new ByteLocation(row, col);
    }

    /**
     * Draw a cell at the specified location.
     */
    private void drawCell(Graphics2D g2, Cell cell, int xpos, int ypos, UserValue userValue) {

        int s = getScale(pieceSize);

        int jittered_xpos = xpos + (int)(Math.random() * 3 - 1);
        int jittered_ypos = ypos + (int)(Math.random() * 3 - 1);
        Font font = new Font(GUIUtil.DEFAULT_FONT_FAMILY, Font.PLAIN, pieceSize >> 1);

        g2.setFont(font);
        g2.setColor( cell.isOriginal() ? CELL_ORIG_BACKGROUND_COLOR : CELL_BACKGROUND_COLOR );
        g2.fillRect( xpos + 1, ypos + 1, pieceSize - 3, pieceSize - 2 );

        if (userValue != null) {
            drawUserValue(g2, userValue, s, xpos, ypos);
        }
        else if (cell.getValue() > 0) {
            g2.setColor( cell.isOriginal() ? CELL_ORIG_TEXT_COLOR : CELL_TEXT_COLOR );
            g2.drawString(ValueConverter.getSymbol(cell.getValue()),
                    jittered_xpos + (int) (0.8 * s), (int) (jittered_ypos + s * 1.7));
        }


        // draw the first 9 numbers in the candidate list, if there are any.
        if (showCandidates) {
            drawCandidates(g2, cell.getCandidates(), xpos, ypos);
        }
    }

    private void drawUserValue(Graphics2D g2, UserValue userValue, int s, int xpos, int ypos) {

        if (userValue.isValidated()) {
            if (userValue.isValid()) {
                g2.setColor(USER_VALUE_CORRECT_COLOR);

            }
            else {
                drawBigX(g2, s, xpos, ypos);
                g2.setColor(USER_VALUE_WRONG_COLOR);
            }
        }
        else {
            g2.setColor(USER_VALUE_COLOR);
        }

        g2.drawString(ValueConverter.getSymbol(userValue.getValue()),
                xpos + (int)(0.8 * s), (int)(ypos + s * 1.7) );
    }

    private void drawBigX(Graphics2D g2, int s, int xpos, int ypos) {
        g2.setColor(BIG_X_COLOR);
        g2.setStroke(BIG_X_STROKE);
        int leftX = xpos + (int)(0.15*s);
        int rightX = xpos + (int)(2.3*s);
        int bottomY = ypos + (int)(2.45*s);
        int topY = ypos + (int)(0.1*s);
        g2.drawLine(leftX, topY, rightX, bottomY);
        g2.drawLine(rightX, topY, leftX, bottomY);
    }

    private void drawCandidates(Graphics g, Candidates candidates, int xpos, int ypos) {

        if (candidates != null ) {
            g.setColor(CANDIDATE_TEXT_COLOR);
            Font candidateFont = new Font("Sans Serif", Font.PLAIN, (pieceSize >> 2) - 2);
            g.setFont(candidateFont);

            drawHints(g, candidates, xpos, ypos, getScale(pieceSize));
        }
    }

    private void drawCurrentFocus(Graphics g, Location focusLocation) {
        int xpos = MARGIN + focusLocation.getCol() * pieceSize;
        int ypos = MARGIN + focusLocation.getRow() * pieceSize;
        g.setColor(CELL_FOCUS_COLOR);
        g.fillRect(xpos, ypos, pieceSize, pieceSize);
    }

    private int getScale(int pieceSize) {
       return (int) (pieceSize * 0.4);
    }
    private void drawHints(Graphics g, Candidates candidates, int x, int y, int scale) {
        int xOffsetLow =  (int) (0.3 * scale);
        int xOffsetMed =  (int) (1.1 * scale);
        int xOffsetHi =  (int) (1.9 * scale);
        int yOffsetLow =  (int) (0.7 * scale);
        int yOffsetMed =  (int) (1.5 * scale);
        int yOffsetHi =  (int) (2.3 * scale);
        int[][] offsets = {
                {xOffsetLow, yOffsetLow},
                {xOffsetMed, yOffsetLow},
                {xOffsetHi, yOffsetLow},
                {xOffsetLow, yOffsetMed},
                {xOffsetMed, yOffsetMed},
                {xOffsetHi, yOffsetMed},
                {xOffsetLow, yOffsetHi},
                {xOffsetMed, yOffsetHi},
                {xOffsetHi, yOffsetHi}
        };

        int ct = 0;
        Iterator<Integer> cit = candidates.iterator();
        while (cit.hasNext() && ct < 9)  {
            g.drawString(ValueConverter.getSymbol(cit.next()), x + offsets[ct][0], y + offsets[ct][1]);
            ct++;
        }
    }

    /**
     * draw the borders around each piece.
     */
    private void drawCellBoundaryGrid(Graphics g, int edgeLen) {

        Graphics2D g2 = (Graphics2D) g;
        int xpos, ypos;

        int rightEdgePos = MARGIN + pieceSize * edgeLen;
        int bottomEdgePos = MARGIN + pieceSize * edgeLen;
        int bigCellLen = (int) Math.sqrt(edgeLen);

        // draw the hatches which delineate the cells
        g.setColor( GRID_COLOR );
        for ( int i = 0; i <= edgeLen; i++ )  //   -----
        {
            ypos = MARGIN + i * pieceSize;
            g2.setStroke((i % bigCellLen == 0) ? BIG_CELL_STROKE : CELL_STROKE);
            g2.drawLine( MARGIN, ypos, rightEdgePos, ypos );
        }
        for ( int i = 0; i <= edgeLen; i++ )  //   ||||
        {
            xpos = MARGIN + i * pieceSize;
            g2.setStroke((i % bigCellLen == 0) ? BIG_CELL_STROKE : CELL_STROKE);
            g2.drawLine( xpos, MARGIN, xpos, bottomEdgePos );
        }
    }
}

File: scala-source/com/barrybecker4/puzzle/sudoku/ui/SudokuRenderer.scala
// Copyright by Barry G. Becker, 2017. Licensed under MIT License: http://www.opensource.org/licenses/MIT
package com.barrybecker4.puzzle.sudoku.ui

import java.awt._

import com.barrybecker4.common.geometry.{ByteLocation, Location}
import com.barrybecker4.puzzle.sudoku.model.ValueConverter
import com.barrybecker4.puzzle.sudoku.model.board.{Board, Cell}
import com.barrybecker4.ui.util.GUIUtil


/**
  * Renders the the sudoku puzzle onscreen.
  *
  * @author Barry Becker
  */
object SudokuRenderer {
  private val MARGIN = 50
  private val CELL_ORIG_TEXT_COLOR = Color.BLACK
  private val CELL_TEXT_COLOR = Color.BLUE
  private val CELL_ORIG_BACKGROUND_COLOR = new Color(215, 225, 238, 200)
  private val CELL_BACKGROUND_COLOR = new Color(245, 245, 255, 100)
  private val CELL_FOCUS_COLOR = new Color(255, 250, 200)
  private val USER_VALUE_COLOR = new Color(155, 5, 40)
  private val USER_VALUE_CORRECT_COLOR = new Color(0, 200, 0)
  private val USER_VALUE_WRONG_COLOR = new Color(255, 10, 0)
  private val BIG_X_COLOR = new Color(255, 0, 0, 50)
  private val BIG_X_STROKE = new BasicStroke(10, BasicStroke.CAP_BUTT, BasicStroke.JOIN_BEVEL)
  private val GRID_COLOR = new Color(10, 0, 100)
  private val TEXT_COLOR = new Color(0, 10, 10)
  private val BACKGROUND_COLOR = new Color(220, 220, 240)
  private val CANDIDATE_TEXT_COLOR = new Color(160, 160, 210)
  private val CELL_STROKE = new BasicStroke(0.5f)
  private val BIG_CELL_STROKE = new BasicStroke(3.0f)
}

class SudokuRenderer(var board: Board) extends CellLocator {
  private var showCandidates: Boolean = false
  private var pieceSize: Int = 0


  def setShowCandidates (show: Boolean) {
    showCandidates = show
  }

  /** This renders the current state of the Slider to the screen. */
  def render(g: Graphics, userEnteredValues: Map[Location, UserValue],
             currentFocusLocation: Location, width: Int, height: Int) {
    val g2: Graphics2D = g.asInstanceOf[Graphics2D]
    val minEdge: Int = Math.min (width, height) - 20 - SudokuRenderer.MARGIN
    pieceSize = minEdge / board.edgeLength
    // erase what's there and redraw.
    g.setColor (SudokuRenderer.BACKGROUND_COLOR)
    g.fillRect (0, 0, width, height)
    g.setColor (SudokuRenderer.TEXT_COLOR)
    g.drawString ("Number of tries: " + board.numIterations, SudokuRenderer.MARGIN, SudokuRenderer.MARGIN - 24)
    val len: Int = board.edgeLength
    var xpos: Int = 0
    var ypos: Int = 0
    if (currentFocusLocation != null) {
      drawCurrentFocus (g, currentFocusLocation)
    }

    for (i <- 0 until len; j <- 0 until len) {
      val loc = (i + 1, j + 1)
      val c: Cell = board.getCell(loc)
      val cands = board.getValues(loc)
      xpos = SudokuRenderer.MARGIN + j * pieceSize
      ypos = SudokuRenderer.MARGIN + i * pieceSize
      drawCell(g2, c, cands, xpos, ypos, userEnteredValues.get(new ByteLocation(i, j)))
    }
    drawCellBoundaryGrid (g, len)
  }

  def getCellCoordinates(point: Point): Location = {
    val row: Int = ((point.getY - SudokuRenderer.MARGIN) / pieceSize).toInt
    val col: Int = ((point.getX - SudokuRenderer.MARGIN) / pieceSize).toInt
    new ByteLocation (row, col)
  }

  /**
    * Draw a cell at the specified location.
    */
  private def drawCell(g2: Graphics2D, cell: Cell, cands: Seq[Int], xpos: Int, ypos: Int, userValue: Option[UserValue]) {
    val s: Int = getScale (pieceSize)
    val jitteredXpos: Int = xpos + (Math.random * 3 - 1).toInt
    val jitteredYpos: Int = ypos + (Math.random * 3 - 1).toInt
    val font: Font = new Font (GUIUtil.DEFAULT_FONT_FAMILY, Font.PLAIN, pieceSize >> 1)
    g2.setFont (font)
    val bg = if (cell.originalValue > 0) SudokuRenderer.CELL_ORIG_BACKGROUND_COLOR else SudokuRenderer.CELL_BACKGROUND_COLOR
    g2.setColor(bg)
    g2.fillRect (xpos + 1, ypos + 1, pieceSize - 3, pieceSize - 2)
    if (userValue.isDefined) {
      drawUserValue (g2, userValue.get, s, xpos, ypos)
    }
    else if (cell.proposedValue > 0) {
      g2.setColor (if (cell.originalValue > 0) SudokuRenderer.CELL_ORIG_TEXT_COLOR
      else SudokuRenderer.CELL_TEXT_COLOR)
      g2.drawString (ValueConverter.getSymbol (cell.proposedValue),
        jitteredXpos + (0.8 * s).toInt, (jitteredYpos + s * 1.7).toInt)
    }
    // draw the first 9 numbers in the candidate list, if there are any.
    if (showCandidates) {
      drawCandidates(g2, cands, xpos, ypos)
    }
  }

  private def drawUserValue (g2: Graphics2D, userValue: UserValue, s: Int, xpos: Int, ypos: Int) {
    if (userValue.isValidated) {
      if (userValue.isValid) {
        g2.setColor(SudokuRenderer.USER_VALUE_CORRECT_COLOR)
      }
      else {
        drawBigX (g2, s, xpos, ypos)
        g2.setColor(SudokuRenderer.USER_VALUE_WRONG_COLOR)
      }
    }
    else {
      g2.setColor(SudokuRenderer.USER_VALUE_COLOR)
    }
    g2.drawString(ValueConverter.getSymbol (userValue.getValue), xpos + (0.8 * s).toInt, (ypos + s * 1.7).toInt)
  }

  private def drawBigX(g2: Graphics2D, s: Int, xpos: Int, ypos: Int) {
    g2.setColor (SudokuRenderer.BIG_X_COLOR)
    g2.setStroke (SudokuRenderer.BIG_X_STROKE)
    val leftX: Int = xpos + (0.15 * s).toInt
    val rightX: Int = xpos + (2.3 * s).toInt
    val bottomY: Int = ypos + (2.45 * s).toInt
    val topY: Int = ypos + (0.1 * s).toInt
    g2.drawLine (leftX, topY, rightX, bottomY)
    g2.drawLine (rightX, topY, leftX, bottomY)
  }

  private def drawCandidates(g: Graphics, candidates: Seq[Int], xpos: Int, ypos: Int) {
    if (candidates.nonEmpty) {
      g.setColor (SudokuRenderer.CANDIDATE_TEXT_COLOR)
      val candidateFont: Font = new Font ("Sans Serif", Font.PLAIN, (pieceSize >> 2) - 2)
      g.setFont (candidateFont)
      drawHints (g, candidates, xpos, ypos, getScale(pieceSize) )
    }
  }

  private def drawCurrentFocus(g: Graphics, focusLocation: Location) {
    val xpos: Int = SudokuRenderer.MARGIN + focusLocation.getCol * pieceSize
    val ypos: Int = SudokuRenderer.MARGIN + focusLocation.getRow * pieceSize
    g.setColor (SudokuRenderer.CELL_FOCUS_COLOR)
    g.fillRect (xpos, ypos, pieceSize, pieceSize)
  }

  private def getScale(pieceSize: Int): Int = (pieceSize * 0.4).toInt

  private def drawHints(g: Graphics, candidates: Seq[Int], x: Int, y: Int, scale: Int) {
    val xOffsetLow: Int = (0.3 * scale).toInt
    val xOffsetMed: Int = (1.1 * scale).toInt
    val xOffsetHi: Int = (1.9 * scale).toInt
    val yOffsetLow: Int = (0.7 * scale).toInt
    val yOffsetMed: Int = (1.5 * scale).toInt
    val yOffsetHi: Int = (2.3 * scale).toInt
    val offsets: Array[Array[Int]] = Array (
      Array (xOffsetLow, yOffsetLow),
      Array (xOffsetMed, yOffsetLow),
      Array (xOffsetHi, yOffsetLow),
      Array (xOffsetLow, yOffsetMed),
      Array (xOffsetMed, yOffsetMed),
      Array (xOffsetHi, yOffsetMed),
      Array (xOffsetLow, yOffsetHi),
      Array (xOffsetMed, yOffsetHi),
      Array (xOffsetHi, yOffsetHi)
    )
    var ct: Int = 0
    for (cand <- candidates if ct < 9) {
      g.drawString(ValueConverter.getSymbol(cand), x + offsets(ct)(0), y + offsets(ct)(1) )
      ct += 1
    }
  }

  /** draw the borders around each piece. */
  private def drawCellBoundaryGrid (g: Graphics, edgeLen: Int) {
    val g2: Graphics2D = g.asInstanceOf[Graphics2D]
    var xpos: Int = 0
    var ypos: Int = 0
    val rightEdgePos: Int = SudokuRenderer.MARGIN + pieceSize * edgeLen
    val bottomEdgePos: Int = SudokuRenderer.MARGIN + pieceSize * edgeLen
    val bigCellLen: Int = Math.sqrt (edgeLen).toInt
    // draw the hatches which delineate the cells
    g.setColor (SudokuRenderer.GRID_COLOR)

    for (i <- 0 to edgeLen) { //   -----
      ypos = SudokuRenderer.MARGIN + i * pieceSize
      g2.setStroke(if (i % bigCellLen == 0) SudokuRenderer.BIG_CELL_STROKE else SudokuRenderer.CELL_STROKE)
      g2.drawLine (SudokuRenderer.MARGIN, ypos, rightEdgePos, ypos)
    }

    for (i <- 0 to edgeLen) { //   ||||
      xpos = SudokuRenderer.MARGIN + i * pieceSize
      g2.setStroke (if (i % bigCellLen == 0) SudokuRenderer.BIG_CELL_STROKE else SudokuRenderer.CELL_STROKE)
      g2.drawLine (xpos, SudokuRenderer.MARGIN, xpos, bottomEdgePos)
    }
  }
}